The data wrangling for my dividend capture strategy for the S&P 500 Dividend Aristocrats capstone project will be in two main parts.  

The first part will involve capturing the company data (not price or dividend data) for the companies that comprise the S&P 500 Dividend Aristocrats index, as well as details for the various index benchmarks that will be used during the analysis.  These benchmarks include the main S&P 500 index, the eleven S&P 500 sector indices, and finally the exchange-traded fund (ETF) that tracks the Dividend Aristocrats index (ticker = NOBL).  

The second part will involve fetching the price, volume, dividend, and stock split data for all of the above.  Please note that indices naturally do not possess dividend or stock split data.

The final output will be saved .csv files for the following:  aristocrats_complete_for_time_series.csv, indices_for_time_series.csv, merged_final_for_time_series.csv

The source for company information for the Aristocrats will be Wikipedia, and the primary source for price/volume/dividend/split data will be the yfinance library for python.  Dividend data from dividendhistory.org will be used to reconcile discrepancies found in the yfinance dividend data.  Dividendhistory.org will also be called upon to provide historical dividends for those aristocrats that have since been delisted.  For those where yfinance no longer supplies historical price data, I will rely on my Norgate paid subscription to supply the historical open, high, low, close, and volume data.

In [1]:
# Start by wrangling the companies that comprise the S&P 500 Dividend Aristocrats.
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
from io import StringIO

# Wikipedia page for S&P 500 Dividend Aristocrats.
aristocrats_url = 'https://en.wikipedia.org/wiki/S%26P_500_Dividend_Aristocrats'

# Add a user-agent header to mimic a browser request.
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
}

# Make the request with headers.
response = requests.get(aristocrats_url, headers=headers)
soup = BeautifulSoup(response.content, 'html.parser')

# Create list of tables from Wikipedia page.
aristocrats_companies = pd.read_html(StringIO(str(soup)))

# Verify that datatype of list is returned
type(aristocrats_companies)

list

In [2]:
# Visually inspect the data scraped from Wikipedia article
print(aristocrats_companies)

[   Ticker symbol                       Company                  Sector
0            AOS                    A.O. Smith             Industrials
1            ABT           Abbott Laboratories             Health Care
2           ABBV                        AbbVie             Health Care
3            AFL                         AFLAC              Financials
4            APD      Air Products & Chemicals               Materials
..           ...                           ...                     ...
64          TROW        T Rowe Price Group Inc              Financials
65           TGT                   Target Corp  Consumer Discretionary
66           GWW                W. W. Grainger             Industrials
67           WMT                  Walmart Inc.        Consumer Staples
68           WST  West Pharmaceutical Services             Health Care

[69 rows x 3 columns],     Year                                              Added  \
0   2025  Erie Indemnity (ERIE), Eversource Energy (ES),... 

In [3]:
# Fetch 1st table from list.  This table lists the current companies in the index.
# Verify DataFrame from 1st table's contents and visually inspect first few records.
aristocrats = aristocrats_companies[0]
print(type(aristocrats))
print(aristocrats.head(20))

<class 'pandas.core.frame.DataFrame'>
   Ticker symbol                    Company                  Sector
0            AOS                 A.O. Smith             Industrials
1            ABT        Abbott Laboratories             Health Care
2           ABBV                     AbbVie             Health Care
3            AFL                      AFLAC              Financials
4            APD   Air Products & Chemicals               Materials
5            ALB      Albemarle Corporation               Materials
6           AMCR                      Amcor               Materials
7            ADM  Archer-Daniels-Midland Co        Consumer Staples
8            ATO          Atmos Energy Corp               Utilities
9            ADP  Automatic Data Processing  Information Technology
10           BDX      Becton Dickinson & Co             Health Care
11           BRO         Brown & Brown Inc.              Financials
12          BF.B     Brown–Forman (class B)        Consumer Staples
13        

In [4]:
# Change Ticker symbol column header to just Ticker.
aristocrats.rename(columns={"Ticker symbol": "Ticker"}, inplace=True)
aristocrats.head()

,Ticker,Company,Sector
0,AOS,A.O. Smith,Industrials
1,ABT,Abbott Laboratories,Health Care
2,ABBV,AbbVie,Health Care
3,AFL,AFLAC,Financials
4,APD,Air Products & Chemicals,Materials


In [5]:
# Fetch 2nd table from list.
# This table is focused on additions and removals of companies over the life of the index.
# Company survivorship of an index is an integral part of the backtesting of any investment strategy.
additions_removals = aristocrats_companies[1]
print(type(additions_removals))
print(additions_removals)

<class 'pandas.core.frame.DataFrame'>
    Year                                              Added  \
0   2025  Erie Indemnity (ERIE), Eversource Energy (ES),...   
1   2024                                Fastenal Co. (FAST)   
2   2023  CH Robinson Worldwide (CHRW), Nordson (NDSN), ...   
3   2022  Brown & Brown, Inc. (BRO) and Church & Dwight ...   
4   2021  IBM (IBM), NextEra Energy (NEE) and West Pharm...   
5   2020  Amcor (AMCR), Atmos Energy (ATO), Realty Incom...   
6   2019  Chubb Limited (CB), People's United Financial ...   
7   2018                                               none   
8   2017                                               none   
9   2016                                               none   
10  2015                                               none   
11  2014                                               none   
12  2013                                               none   
13  2012  AT&T (T), Colgate-Palmolive (CL), Franklin Tem...   
14  2011  Ecolab 

In [6]:
# I intend to start the machine learning for time series 
# with year 2013 as NOBL was launched in 2013; therefore, 
# I only require the additions and removals since 2013 (inclusive).
additions_removals = additions_removals[additions_removals['Year'] >= 2013]
additions_removals

,Year,Added,Removed,Reference(s)
0,2025,"Erie Indemnity (ERIE), Eversource Energy (ES),...",none,[7]
1,2024,Fastenal Co. (FAST),Walgreens Boots Alliance Inc. (WBA) and the 3M...,[8][9]
2,2023,"CH Robinson Worldwide (CHRW), Nordson (NDSN), ...",VF Corporation (VFC),[10][11][12]
3,2022,"Brown & Brown, Inc. (BRO) and Church & Dwight ...","AT&T Inc. (T); People's United Financial, Inc....",[13][14][15]
4,2021,"IBM (IBM), NextEra Energy (NEE) and West Pharm...","Raytheon (RTX), Carrier Global (CARR), Otis Wo...",[16]
5,2020,"Amcor (AMCR), Atmos Energy (ATO), Realty Incom...",Ross Stores (ROST) and Helmerich & Payne (HP).,[17]
6,2019,"Chubb Limited (CB), People's United Financial ...",none,[18]
7,2018,none,none,NaN
8,2017,none,none,NaN
9,2016,none,Chubb Corp (CB) (upon acquisition by ACE Limit...,NaN


In [7]:
# Parse the tickers from the 'Added' column and include in new 'Added_Tickers' column.
additions_removals = additions_removals.copy()
additions_removals['Added_Tickers'] = additions_removals.loc[:, 'Added'].str.findall(r'\(([A-Z\s]+)\)')

# Parse the tickers from the 'Removed' column and include in new 'Removed_Tickers' column.
additions_removals = additions_removals.copy()
additions_removals['Removed_Tickers'] = additions_removals.loc[:, 'Removed'].str.findall(r'\(([A-Z\s]+)\)')
additions_removals

,Year,Added,Removed,Reference(s),Added_Tickers,Removed_Tickers
0,2025,"Erie Indemnity (ERIE), Eversource Energy (ES),...",none,[7],"[ERIE, ES, RGLD, FDS]",[]
1,2024,Fastenal Co. (FAST),Walgreens Boots Alliance Inc. (WBA) and the 3M...,[8][9],[FAST],"[WBA, MMM]"
2,2023,"CH Robinson Worldwide (CHRW), Nordson (NDSN), ...",VF Corporation (VFC),[10][11][12],"[CHRW, NDSN, SJM, KVUE]",[VFC]
3,2022,"Brown & Brown, Inc. (BRO) and Church & Dwight ...","AT&T Inc. (T); People's United Financial, Inc....",[13][14][15],"[BRO, CHD]","[T, PBCT]"
4,2021,"IBM (IBM), NextEra Energy (NEE) and West Pharm...","Raytheon (RTX), Carrier Global (CARR), Otis Wo...",[16],"[IBM, NEE, WST]","[RTX, CARR, OTIS, SYK, LEG]"
5,2020,"Amcor (AMCR), Atmos Energy (ATO), Realty Incom...",Ross Stores (ROST) and Helmerich & Payne (HP).,[17],"[AMCR, ATO, O, ESS, ROST, ALB, EXPD, CARR, OTIS]","[ROST, HP]"
6,2019,"Chubb Limited (CB), People's United Financial ...",none,[18],"[CB, PBCT, CAT, UTX]",[]
7,2018,none,none,NaN,[],[]
8,2017,none,none,NaN,[],[]
9,2016,none,Chubb Corp (CB) (upon acquisition by ACE Limit...,NaN,[],"[CB, ACE]"


In [8]:
# The Added, Removed, and Reference(s) columns are no longer required at this point.
additions_removals = additions_removals.drop(columns=['Added', 'Removed', 'Reference(s)'])
additions_removals

,Year,Added_Tickers,Removed_Tickers
0,2025,"[ERIE, ES, RGLD, FDS]",[]
1,2024,[FAST],"[WBA, MMM]"
2,2023,"[CHRW, NDSN, SJM, KVUE]",[VFC]
3,2022,"[BRO, CHD]","[T, PBCT]"
4,2021,"[IBM, NEE, WST]","[RTX, CARR, OTIS, SYK, LEG]"
5,2020,"[AMCR, ATO, O, ESS, ROST, ALB, EXPD, CARR, OTIS]","[ROST, HP]"
6,2019,"[CB, PBCT, CAT, UTX]",[]
7,2018,[],[]
8,2017,[],[]
9,2016,[],"[CB, ACE]"


In [9]:
# Explode the list of tickers in the Added_Tickers column.
# Insert the content into new DataFrame along with Year value.
additions_filtered = (additions_removals[additions_removals['Added_Tickers'].apply(
    lambda x: x is not None and hasattr(x, '__len__') and len(x) > 0)]  # Changed isinstance check
    .explode('Added_Tickers')
    .rename(columns={'Added_Tickers': 'Ticker', 'Year': 'Year_Added'})
    [['Year_Added', 'Ticker']]
    .reset_index(drop=True))


# Explode the list of tickers in the Removed_Tickers column.
# Insert the content into new DataFrame along with Year value.
removals_filtered = (additions_removals[additions_removals['Removed_Tickers'].apply(
    lambda x: x is not None and hasattr(x, '__len__') and len(x) > 0)]  # Changed isinstance check
    .explode('Removed_Tickers')
    .rename(columns={'Removed_Tickers': 'Ticker', 'Year': 'Year_Removed'})
    [['Year_Removed', 'Ticker']]
    .reset_index(drop=True))

# Merge the two above DataFrames into a single source of truth.
combined_filtered = pd.merge(additions_filtered, 
                            removals_filtered, 
                            on='Ticker', 
                            how='outer')

# Reorder the columns.
col_reorder = ['Ticker', 'Year_Added', 'Year_Removed']
combined_filtered = combined_filtered[col_reorder]

# Examine the final result.
combined_filtered

,Ticker,Year_Added,Year_Removed
0,ACE,NaN,2016.0
1,ALB,2020.0,NaN
2,AMCR,2020.0,NaN
3,ATO,2020.0,NaN
4,BMS,NaN,2014.0
5,BRO,2022.0,NaN
6,CARR,2020.0,2021.0
7,CAT,2019.0,NaN
8,CB,2019.0,2016.0
9,CHD,2022.0,NaN


In [10]:
# Since I am having all additions and removals occur on January 1st,
# any ticker that is added and removed in same year should be omitted.
combined_filtered = combined_filtered[combined_filtered['Year_Added'] != combined_filtered['Year_Removed']]
combined_filtered.reset_index(drop=True, inplace=True) 
combined_filtered

,Ticker,Year_Added,Year_Removed
0,ACE,NaN,2016.0
1,ALB,2020.0,NaN
2,AMCR,2020.0,NaN
3,ATO,2020.0,NaN
4,BMS,NaN,2014.0
5,BRO,2022.0,NaN
6,CARR,2020.0,2021.0
7,CAT,2019.0,NaN
8,CB,2019.0,2016.0
9,CHD,2022.0,NaN


In [11]:
# Merge the aristocrats and combined_filters DataFrames.
aristocrats_complete = pd.merge(aristocrats, 
                            combined_filtered, 
                            on='Ticker', 
                            how='outer')
aristocrats_complete.head(20)

,Ticker,Company,Sector,Year_Added,Year_Removed
0,ABBV,AbbVie,Health Care,NaN,NaN
1,ABT,Abbott Laboratories,Health Care,NaN,NaN
2,ACE,NaN,NaN,NaN,2016.0
3,ADM,Archer-Daniels-Midland Co,Consumer Staples,NaN,NaN
4,ADP,Automatic Data Processing,Information Technology,NaN,NaN
5,AFL,AFLAC,Financials,NaN,NaN
6,ALB,Albemarle Corporation,Materials,2020.0,NaN
7,AMCR,Amcor,Materials,2020.0,NaN
8,AOS,A.O. Smith,Industrials,NaN,NaN
9,APD,Air Products & Chemicals,Materials,NaN,NaN


In [12]:
# For those records where Year_Added is NaN, insert 2013 (start date for
# time series machine learning).
aristocrats_complete['Year_Added'] = aristocrats_complete['Year_Added'].fillna(2013)

# For those records where Year_Removed is NaN, insert 2030 (arbitrary year
# in near future).
aristocrats_complete['Year_Removed'] = aristocrats_complete['Year_Removed'].fillna(2030)

aristocrats_complete.head(20)

,Ticker,Company,Sector,Year_Added,Year_Removed
0,ABBV,AbbVie,Health Care,2013.0,2030.0
1,ABT,Abbott Laboratories,Health Care,2013.0,2030.0
2,ACE,NaN,NaN,2013.0,2016.0
3,ADM,Archer-Daniels-Midland Co,Consumer Staples,2013.0,2030.0
4,ADP,Automatic Data Processing,Information Technology,2013.0,2030.0
5,AFL,AFLAC,Financials,2013.0,2030.0
6,ALB,Albemarle Corporation,Materials,2020.0,2030.0
7,AMCR,Amcor,Materials,2020.0,2030.0
8,AOS,A.O. Smith,Industrials,2013.0,2030.0
9,APD,Air Products & Chemicals,Materials,2013.0,2030.0


**Note:**  Yfinance does not currently accept the period (.) delimiter to distinguish share class for the ticker symbol, and any call to yfinance using a ticker that has this delimiter (e.g., "BF.B") will fail.  Yfinance utilizes the hyphen char (-) instead.  

In [13]:
# Identify any tickers that have period (.) char as substring and  
# replace with hyphen (-) char.
aristocrats_complete['Ticker'] = aristocrats_complete['Ticker'].str.\
                                    replace('.', '-', regex=False)
aristocrats_complete.head(20)

,Ticker,Company,Sector,Year_Added,Year_Removed
0,ABBV,AbbVie,Health Care,2013.0,2030.0
1,ABT,Abbott Laboratories,Health Care,2013.0,2030.0
2,ACE,NaN,NaN,2013.0,2016.0
3,ADM,Archer-Daniels-Midland Co,Consumer Staples,2013.0,2030.0
4,ADP,Automatic Data Processing,Information Technology,2013.0,2030.0
5,AFL,AFLAC,Financials,2013.0,2030.0
6,ALB,Albemarle Corporation,Materials,2020.0,2030.0
7,AMCR,Amcor,Materials,2020.0,2030.0
8,AOS,A.O. Smith,Industrials,2013.0,2030.0
9,APD,Air Products & Chemicals,Materials,2013.0,2030.0


There are several tickers in the DataFrame with missing Company and Sector.  Use yfinance to fetch these values.  I suspect that some of these tickers will return Company=None and Sector=None as they have since been delisted from the exchanges.

In [14]:
import yfinance as yf
import time

for index, row in aristocrats_complete.iterrows():
    if not pd.isna(row['Company']):
        continue
    try:
        info = yf.Ticker(row['Ticker']).info
        aristocrats_complete.at[index, 'Sector'] = info.get('sector')
        aristocrats_complete.at[index, 'Company'] = info.get('longName')
        time.sleep(0.5)
    except Exception as e:
        print(f"Error fetching {row['Ticker']}: {e}")

HTTP Error 404: {"quoteSummary":{"result":null,"error":{"code":"Not Found","description":"Quote not found for symbol: WBA"}}}


**Note:** The 404 error returned for WBA indicates that WBA is the first of a possible handful of symbols that have been delisted from the exchanges.  I'll be watching for these in subsequent cells.

In [15]:
# Quick look at the results.
print(aristocrats_complete.head(15))
print(aristocrats_complete.tail(15))

   Ticker                    Company                  Sector  Year_Added  \
0    ABBV                     AbbVie             Health Care      2013.0   
1     ABT        Abbott Laboratories             Health Care      2013.0   
2     ACE                       None                    None      2013.0   
3     ADM  Archer-Daniels-Midland Co        Consumer Staples      2013.0   
4     ADP  Automatic Data Processing  Information Technology      2013.0   
5     AFL                      AFLAC              Financials      2013.0   
6     ALB      Albemarle Corporation               Materials      2020.0   
7    AMCR                      Amcor               Materials      2020.0   
8     AOS                 A.O. Smith             Industrials      2013.0   
9     APD   Air Products & Chemicals               Materials      2013.0   
10    ATO          Atmos Energy Corp               Utilities      2020.0   
11    BDX      Becton Dickinson & Co             Health Care      2013.0   
12    BEN   

In [16]:
# Upon further inspection of the Wikipedia article, ticker ACE was only mentioned as the 
# company that acquired Chubb Corp (CB) in 2016.  In light of this, ticker ACE can be removed.
aristocrats_complete = aristocrats_complete[aristocrats_complete.Ticker != 'ACE'].reset_index(drop=True)
aristocrats_complete.head()

,Ticker,Company,Sector,Year_Added,Year_Removed
0,ABBV,AbbVie,Health Care,2013.0,2030.0
1,ABT,Abbott Laboratories,Health Care,2013.0,2030.0
2,ADM,Archer-Daniels-Midland Co,Consumer Staples,2013.0,2030.0
3,ADP,Automatic Data Processing,Information Technology,2013.0,2030.0
4,AFL,AFLAC,Financials,2013.0,2030.0


In [17]:
# Return those tickers that I suspect as being delisted.
# I suspect Sector=None will be returned for all.
delisted = aristocrats_complete[aristocrats_complete['Company'].isna()]
delisted

,Ticker,Company,Sector,Year_Added,Year_Removed
35,FDO,None,None,2013.0,2015.0
61,PBCT,None,None,2019.0,2022.0
71,SIAL,None,None,2013.0,2015.0
80,UTX,None,None,2019.0,2030.0
82,WBA,None,None,2013.0,2024.0


In [18]:
# Need to confirm my suspicions by attempting to fetch price data for these
for symbol in delisted['Ticker']:
    price_hist = yf.Ticker(symbol).history(period='max', auto_adjust=False)
    len(price_hist)

$FDO: possibly delisted; no price data found  (1d 1927-08-03 -> 2026-07-09)
$PBCT: possibly delisted; no timezone found
$SIAL: possibly delisted; no price data found  (1d 1927-08-03 -> 2026-07-09)
$UTX: possibly delisted; no timezone found
$WBA: possibly delisted; no timezone found


For symbols BMS, FDO, PBCT, SIAL, and WBA, I can retrieve dividend history from dividendhistory.org, and I can retrieve the historical price data from my Norgate subscription.  Independent research uncovered that UTX (United Technologies) was removed from the S&P 500 Dividend Aristocrats index prior to the market open on April 6, 2020.  In April 2020 the company merged with Raytheon (RTN), and the merged company now trades as Raytheon Technologies under the symbol RTX.  The upshot is that the price and dividend history for UTX is already encapsulated in RTX, therefore, there is no need to keep symbol UTX in the aristocrats_complete DataFrame.

In [19]:
aristocrats_complete.drop(
    aristocrats_complete[aristocrats_complete['Ticker'] == 'UTX'].index,
    inplace=True
)

# Since I'm removing the row for UTX, I need to adjust the Year_Added to 2019 for RTX.
aristocrats_complete.loc[aristocrats_complete['Ticker'] == 'RTX', 'Year_Added'] = 2019

# Reset index and do sanity check.
aristocrats_complete.reset_index(drop=True, inplace=True) 
aristocrats_complete.tail(20)

,Ticker,Company,Sector,Year_Added,Year_Removed
65,PNR,Pentair,Industrials,2013.0,2030.0
66,PPG,PPG Industries,Materials,2013.0,2030.0
67,RGLD,"Royal Gold, Inc.",Basic Materials,2025.0,2030.0
68,ROP,Roper Technologies,Industrials,2013.0,2030.0
69,RTX,RTX Corporation,Industrials,2019.0,2021.0
70,SHW,Sherwin-Williams,Materials,2013.0,2030.0
71,SIAL,None,None,2013.0,2015.0
72,SJM,The J. M. Smucker Company,Consumer Staples,2023.0,2030.0
73,SPGI,S&P Global Inc,Financials,2013.0,2030.0
74,SWK,Stanley Black & Decker,Industrials,2013.0,2030.0


In [20]:
# Cast the Year_Added and Year_Removed columns to integer.
aristocrats_complete['Year_Added'] = aristocrats_complete['Year_Added'].astype('int')
aristocrats_complete['Year_Removed'] = aristocrats_complete['Year_Removed'].astype('int')
aristocrats_complete.dtypes

Ticker          object
Company         object
Sector          object
Year_Added       int64
Year_Removed     int64
dtype: object

In [21]:
# Determine any records where Sector value is missing
aristocrats_complete[aristocrats_complete['Sector'].isna()]

,Ticker,Company,Sector,Year_Added,Year_Removed
13,BMS,"Bemis Company, Inc.",None,2013,2014
35,FDO,None,None,2013,2015
61,PBCT,None,None,2019,2022
71,SIAL,None,None,2013,2015
81,WBA,None,None,2013,2024


In [22]:
# Some research on Google showed that Bemis Company Inc. was the Materials 
# sector of the economy
aristocrats_complete.loc[aristocrats_complete['Ticker'] == 'BMS', 'Sector'] = 'Materials'

In [23]:
# Need to insert data for Company and Sector for the four remaining ticker symbols.
company_map = {
    'FDO': 'Family Dollar Stores',
    'PBCT': "People's United Financial",
    'SIAL': 'Sigma-Aldrich',
    'WBA': 'Walgreens Boots Alliance',
}

sector_map = {
    'FDO': 'Consumer Discretionary',
    'PBCT': 'Financials',
    'SIAL': 'Materials',
    'WBA': 'Consumer Staples',
}

mask = aristocrats_complete['Ticker'].isin(company_map.keys())

aristocrats_complete.loc[mask, 'Company'] = aristocrats_complete.loc[mask, 'Ticker'].map(company_map)
aristocrats_complete.loc[mask, 'Sector'] = aristocrats_complete.loc[mask, 'Ticker'].map(sector_map)

# Quick sanity check
aristocrats_complete.tail(25)

,Ticker,Company,Sector,Year_Added,Year_Removed
60,OTIS,Otis Worldwide Corporation,Industrials,2020,2021
61,PBCT,People's United Financial,Financials,2019,2022
62,PBI,Pitney Bowes Inc.,Industrials,2013,2013
63,PEP,PepsiCo,Consumer Staples,2013,2030
64,PG,Procter & Gamble,Consumer Staples,2013,2030
65,PNR,Pentair,Industrials,2013,2030
66,PPG,PPG Industries,Materials,2013,2030
67,RGLD,"Royal Gold, Inc.",Basic Materials,2025,2030
68,ROP,Roper Technologies,Industrials,2013,2030
69,RTX,RTX Corporation,Industrials,2019,2021


In [24]:
# The above visual inspection shows that some of the Sector values appear to be 
# different depending from where they were fetched (Wikipedia versus yfinance).  
# Need to double check this.
aristocrats_complete['Sector'].unique()

array(['Health Care', 'Consumer Staples', 'Information Technology',
       'Financials', 'Materials', 'Industrials', 'Utilities', 'Energy',
       'Real Estate', 'Consumer Discretionary', 'Consumer Cyclical',
       'Basic Materials', 'Healthcare', 'Communication Services'],
      dtype=object)

In [25]:
# For those Sector verbiages showing discrepancies, realign using intended verbiages. 
# These same verbiages will be used when creating the dictionary of indices further
# down in this Notebook.
aristocrats_complete['Sector'] = aristocrats_complete['Sector'].replace('Healthcare', 'Health Care')
aristocrats_complete['Sector'] = aristocrats_complete['Sector'].replace('Basic Materials', 'Materials')
aristocrats_complete['Sector'] = aristocrats_complete['Sector'].replace('Consumer Cyclical', 'Consumer Discretionary')
aristocrats_complete['Sector'].unique()

# Quick visual inspection
print(aristocrats_complete.tail(15))
print(aristocrats_complete['Sector'].unique())

   Ticker                       Company                  Sector  Year_Added  \
70    SHW              Sherwin-Williams               Materials        2013   
71   SIAL                 Sigma-Aldrich               Materials        2013   
72    SJM     The J. M. Smucker Company        Consumer Staples        2023   
73   SPGI                S&P Global Inc              Financials        2013   
74    SWK        Stanley Black & Decker             Industrials        2013   
75    SYK           Stryker Corporation             Health Care        2013   
76    SYY                    Sysco Corp        Consumer Staples        2013   
77      T                     AT&T Inc.  Communication Services        2013   
78    TGT                   Target Corp  Consumer Discretionary        2013   
79   TROW        T Rowe Price Group Inc              Financials        2013   
80    VFC              V.F. Corporation  Consumer Discretionary        2013   
81    WBA      Walgreens Boots Alliance        Consu

In [26]:
# Iterate through the aristocrats_complete DataFrame and fetch
# historical price and dividend data since 1/1/2013.
# Setting end date to 7/1/2026 will ensure that price data
# for 6/30 is returned.
stock_symbols = aristocrats_complete['Ticker']
stock_data = {
    symbol: yf.Ticker(symbol).history(start='2013-01-01', end='2026-07-01', auto_adjust=False) for symbol in stock_symbols
}
print("Done")

# Quick look at dictionary of market data
print(stock_data)

$BMS: possibly delisted; no price data found  (1d 2013-01-01 -> 2026-07-01)
$FDO: possibly delisted; no price data found  (1d 2013-01-01 -> 2026-07-01)
$PBCT: possibly delisted; no timezone found
$SIAL: possibly delisted; no price data found  (1d 2013-01-01 -> 2026-07-01)
$WBA: possibly delisted; no timezone found


Done
{'ABBV':                                  Open        High         Low       Close  \
Date                                                                        
2013-01-02 00:00:00-05:00   34.919998   35.400002   34.099998   35.119999   
2013-01-03 00:00:00-05:00   35.000000   35.000000   34.160000   34.830002   
2013-01-04 00:00:00-05:00   34.619999   34.889999   34.250000   34.389999   
2013-01-07 00:00:00-05:00   34.150002   35.450001   34.150002   34.459999   
2013-01-08 00:00:00-05:00   34.290001   34.639999   33.360001   33.709999   
...                               ...         ...         ...         ...   
2026-06-24 00:00:00-04:00  236.350006  237.360001  232.850006  234.889999   
2026-06-25 00:00:00-04:00  233.660004  244.610001  233.660004  243.139999   
2026-06-26 00:00:00-04:00  245.259995  253.350006  244.660004  253.350006   
2026-06-29 00:00:00-04:00  253.050003  255.990005  249.199997  254.309998   
2026-06-30 00:00:00-04:00  255.000000  255.500000  250.100006 

Unfortunately yfinance doesn't have a way to return the S&P 500 sector indices as a list; therefore, I will need to manually create a dictionary for the relevant general market and sector indices, as well as the NOBL exchange-traded fund (ETF) benchark that will be used for the time series machine learning.

In [27]:
indices_dict = {
    'Ticker': ['NOBL', 
                      '^GSPC',
					  '^SP500-15', 
					  '^SP500-50', 
					  '^GSPE', 
					  '^SP500-40', 
					  '^SP500-20', 
					  '^SP500-45', 
					  '^SP500-30', 
					  '^SP500-60',
					  '^SP500-55',
					  '^SP500-35',
					  '^SP500-25'],
    'Name': ['S&P 500 Dividend Aristocrats ETF',
             'S&P 500 Index',
	         'S&P 500 Materials Sector Index',
			 'S&P 500 Communication Sector Index',
			 'S&P 500 Energy Sector Index',
			 'S&P 500 Financials Sector Index',
			 'S&P 500 Industrial Sector Index',
			 'S&P 500 Technology Sector Index',
			 'S&P 500 Consumer Staples Sector Index',
			 'S&P 500 Real Estate Sector Index',
			 'S&P 500 Utilities Sector Index',
			 'S&P 500 Health Care Sector Index',
			 'S&P 500 Consumer Discretionary Sector Index'],
    'Type': ['ETF',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index',
             'Index'],
    'Sector': ['NA', 
               'NA',
	           'Materials', 
			   'Communication Services',
			   'Energy',
			   'Financials',
			   'Industrials',
			   'Information Technology',
			   'Consumer Staples',
			   'Real Estate',
			   'Utilities',
			   'Health Care',
			   'Consumer Discretionary']
}

# Convert to DataFrame
indices = pd.DataFrame(indices_dict)
indices

,Ticker,Name,Type,Sector
0,NOBL,S&P 500 Dividend Aristocrats ETF,ETF,NA
1,^GSPC,S&P 500 Index,Index,NA
2,^SP500-15,S&P 500 Materials Sector Index,Index,Materials
3,^SP500-50,S&P 500 Communication Sector Index,Index,Communication Services
4,^GSPE,S&P 500 Energy Sector Index,Index,Energy
5,^SP500-40,S&P 500 Financials Sector Index,Index,Financials
6,^SP500-20,S&P 500 Industrial Sector Index,Index,Industrials
7,^SP500-45,S&P 500 Technology Sector Index,Index,Information Technology
8,^SP500-30,S&P 500 Consumer Staples Sector Index,Index,Consumer Staples
9,^SP500-60,S&P 500 Real Estate Sector Index,Index,Real Estate


In [28]:
# Iterate through the indices DataFrame and fetch
# historical price and dividend data since 1/1/2013.
index_symbols = indices['Ticker']
index_data = {
    symbol: yf.Ticker(symbol).history(start='2013-01-01', end='2026-07-01', auto_adjust=False) for symbol in index_symbols
}
print("Done")

# Quick look at dictionary of index data
print(index_data)

Done
{'NOBL':                                 Open       High        Low      Close  \
Date                                                                    
2013-10-10 00:00:00-04:00  20.200001  20.459999  20.200001  20.375000   
2013-10-11 00:00:00-04:00  20.745001  20.745001  20.350000  20.520000   
2013-10-14 00:00:00-04:00  20.424999  20.605000  20.405001  20.590000   
2013-10-15 00:00:00-04:00  20.555000  20.559999  20.415001  20.434999   
2013-10-16 00:00:00-04:00  20.605000  20.660000  20.495001  20.639999   
...                              ...        ...        ...        ...   
2026-06-24 00:00:00-04:00  54.860001  55.560001  54.849998  55.369999   
2026-06-25 00:00:00-04:00  55.389999  56.349998  55.389999  55.810001   
2026-06-26 00:00:00-04:00  55.820000  56.459999  55.750000  56.290001   
2026-06-29 00:00:00-04:00  56.369999  56.419998  56.070000  56.349998   
2026-06-30 00:00:00-04:00  56.450001  56.500000  56.110001  56.160000   

                           Adj Close

In [29]:
# Verify that I have dictionaries for both stock and index data.
print(type(stock_data))
print(type(index_data))

<class 'dict'>
<class 'dict'>


In [30]:
# Create function to standardize price and dividend data for all asset classes
# (stocks, ETFs, and indexes) that will be used for the time series machine learning.
def standardize_data(data_dict):
    standardized = []
    for symbol, df in data_dict.items():
        if not df.empty:
            df = df[["Open", "High", "Low", "Close", "Adj Close", "Volume", "Dividends", "Stock Splits"]]
            df = df.dropna()
            for col in df.columns:
                if df[col].dtype == 'object':
                    df[col] = df[col].astype(str).str.replace(r'[^\d.-]', '', regex=True)
            df = df.astype(float)
            df["Symbol"] = symbol
            standardized.append(df)
    return standardized

# Standardize the data from both DataFrames, then concatenate.
market_data = pd.concat(
    standardize_data(stock_data) + standardize_data(index_data)
).reset_index()

# Quick sanity check of resulted DataFrame
print(len(market_data))
print(market_data.head())
print(market_data.tail())

309098
                       Date       Open       High        Low      Close  \
0 2013-01-02 00:00:00-05:00  34.919998  35.400002  34.099998  35.119999   
1 2013-01-03 00:00:00-05:00  35.000000  35.000000  34.160000  34.830002   
2 2013-01-04 00:00:00-05:00  34.619999  34.889999  34.250000  34.389999   
3 2013-01-07 00:00:00-05:00  34.150002  35.450001  34.150002  34.459999   
4 2013-01-08 00:00:00-05:00  34.290001  34.639999  33.360001  33.709999   

   Adj Close      Volume  Dividends  Stock Splits Symbol  
0  20.561069  13767900.0        0.0           0.0   ABBV  
1  20.391294  16739300.0        0.0           0.0   ABBV  
2  20.133688  21372100.0        0.0           0.0   ABBV  
3  20.174671  17897100.0        0.0           0.0   ABBV  
4  19.735582  17863300.0        0.0           0.0   ABBV  
                            Date         Open         High          Low  \
309093 2026-06-24 00:00:00-04:00  1848.339966  1898.800049  1848.339966   
309094 2026-06-25 00:00:00-04:00  1852

The Adj Close column can be dropped since it's really not needed.  Adj Close reflects the close adjusted for dividends, but my machine learning requires the price to be raw (that is adjusted only for splits, not for dividends).  As such, it is acceptable to drop this column.

In [31]:
market_data = market_data.drop(columns=['Adj Close'])

In [32]:
# Normalize the values in the Date column to ensure compatibility with future
# pandas operations for time series.
market_data['Date'] = market_data['Date'].dt.tz_localize(None)
market_data.head()

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits,Symbol
0,2013-01-02,34.919998,35.400002,34.099998,35.119999,13767900.0,0.0,0.0,ABBV
1,2013-01-03,35.000000,35.000000,34.160000,34.830002,16739300.0,0.0,0.0,ABBV
2,2013-01-04,34.619999,34.889999,34.250000,34.389999,21372100.0,0.0,0.0,ABBV
3,2013-01-07,34.150002,35.450001,34.150002,34.459999,17897100.0,0.0,0.0,ABBV
4,2013-01-08,34.290001,34.639999,33.360001,33.709999,17863300.0,0.0,0.0,ABBV


In [33]:
# Check for NaN, None, or NaT values
ohlcv = ["Open", "High", "Low", "Close", "Volume"]
nan_data = market_data[market_data[ohlcv].isna().any(axis=1)]
print(len(nan_data))
print(nan_data.head())
print(nan_data.tail())

0
Empty DataFrame
Columns: [Date, Open, High, Low, Close, Volume, Dividends, Stock Splits, Symbol]
Index: []
Empty DataFrame
Columns: [Date, Open, High, Low, Close, Volume, Dividends, Stock Splits, Symbol]
Index: []


In [34]:
# Check zero (0) values for Open, High, Low, and Close for the indices.
# Zero volume is immaterial since I only need price data for the
# indices for my analysis.
ohlc = ["Open", "High", "Low", "Close"]
data_indices = market_data[market_data['Symbol'].str.startswith('^')]
zero_data_indices = data_indices[(data_indices[ohlc] == 0).any(axis=1)]
print(len(zero_data_indices))

0


In [35]:
# Check zero (0) values for Open, High, Low, Close, and Volume for stocks
# and NOBL.  Volume is important here since I will incorporate volume data
# in my machine learning.
ohlcv = ["Open", "High", "Low", "Close", "Volume"]
data_stocks= market_data[~market_data['Symbol'].str.startswith('^')]
zero_data_stocks = data_stocks[(data_stocks[ohlcv] == 0).any(axis=1)]
print(len(zero_data_stocks))

1225


In [36]:
# Let's break this number down by ticker symbols.
print(zero_data_stocks['Symbol'].value_counts())

Symbol
AMCR    1223
CHD        1
ERIE       1
Name: count, dtype: int64


In [37]:
# Since there are only singleton recoreds for CHD and ERIE, let's look at these first.
zero_data_stocks[zero_data_stocks['Symbol'].isin(['CHD', 'ERIE'])]

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits,Symbol
61020,2019-12-31,70.349998,70.349998,70.349998,70.349998,0.0,0.0,0.0,CHD
96822,2013-12-13,70.279999,70.279999,70.279999,70.279999,0.0,0.0,0.0,ERIE


For CHD there was zero volume for 12/31/2019, but the company was first added to the index in 2022.  For ERIE there was zero volume for 12/13/2013, but ERIE was added in 2025.  In light of this, these records can be deleted from the DataFrame.

In [38]:
# Now let's check the details for AMCR.
zero_data_amcr = zero_data_stocks[zero_data_stocks['Symbol'] == 'AMCR']
print(zero_data_amcr.head())
print(zero_data_amcr.tail())

            Date       Open       High        Low      Close  Volume  \
20358 2013-01-02  41.849998  41.849998  41.849998  41.849998     0.0   
20359 2013-01-03  41.849998  41.849998  41.849998  41.849998     0.0   
20360 2013-01-04  41.849998  41.849998  41.849998  41.849998     0.0   
20362 2013-01-08  43.299999  43.299999  43.299999  43.299999     0.0   
20363 2013-01-09  43.299999  43.299999  43.299999  43.299999     0.0   

       Dividends  Stock Splits Symbol  
20358        0.0           0.0   AMCR  
20359        0.0           0.0   AMCR  
20360        0.0           0.0   AMCR  
20362        0.0           0.0   AMCR  
20363        0.0           0.0   AMCR  
            Date   Open   High    Low  Close  Volume  Dividends  Stock Splits  \
21973 2019-06-04  55.75  55.75  55.75  55.75     0.0        0.0           0.0   
21974 2019-06-05  55.75  55.75  55.75  55.75     0.0        0.0           0.0   
21975 2019-06-06  55.75  55.75  55.75  55.75     0.0        0.0           0.0   
219

The records span multiple years, and the very last record is in 2019.  AMCR was not added to the index until 2020, so it will be OK to delete these records.

In [39]:
# Drop zero (0) records for AMCR, CHD, and ERIE
market_data.drop(zero_data_stocks.index, inplace=True)
market_data.reset_index(drop=True, inplace=True) 
print(len(market_data))

307873


In [40]:
# Check for the existence of duped records
duplicate_count = market_data.duplicated(subset=['Date', 'Symbol']).sum()
print(duplicate_count)

0


In [41]:
# Need to bring in historical price data for delisted symbols BMS, FDO, PBCT, SIAL, and WBA 
# from my Norgate subscription.
files_delisted = ['BMS-201906.csv', 'FDO-201507.csv', 'PBCT-202204.csv', 'SIAL-201511.csv', 'WBA-202508.csv']
tickers_delisted = ['BMS', 'FDO', 'PBCT', 'SIAL', 'WBA'] 

cols_to_use = ['Date', 'Open', 'High', 'Low', 'Close', 'Volume']

dfs = []
for file, symbol in zip(files_delisted, tickers_delisted):
    temp_df = pd.read_csv(file, usecols=cols_to_use)
    temp_df['Symbol'] = symbol
    dfs.append(temp_df)

delisted_market_data = pd.concat(dfs, ignore_index=True)

# Quick inspection of the contents.
print(delisted_market_data.head())
print(delisted_market_data.tail())

       Date      Open      High       Low     Close  Volume Symbol
0  19850102  1.664063  1.671875  1.664063  1.671875   17600    BMS
1  19850103  1.656250  1.656250  1.640625  1.640625   27200    BMS
2  19850107  1.640625  1.671875  1.640625  1.640625  414400    BMS
3  19850108  1.640625  1.648438  1.632813  1.632813    9600    BMS
4  19850109  1.625000  1.625000  1.609375  1.609375   16000    BMS
           Date    Open    High    Low  Close     Volume Symbol
42892  20250821  12.148  12.250  12.12  12.19   13097674    WBA
42893  20250822  12.200  12.220  12.07  12.07    9178662    WBA
42894  20250825  12.080  12.175  11.87  11.97   35998112    WBA
42895  20250826  11.940  11.995  11.88  11.92   90375656    WBA
42896  20250827  11.915  12.110  11.90  11.98  452081056    WBA


In [42]:
# Need to format Date column in order to facilitate merge into market_data DataFrame.
delisted_market_data['Date'] = pd.to_datetime(delisted_market_data['Date'], format='%Y%m%d').dt.strftime('%-m/%-d/%Y')
delisted_market_data['Date'] = pd.to_datetime(delisted_market_data['Date']).dt.tz_localize(None)
print(delisted_market_data.dtypes)
print(delisted_market_data.head())

Date      datetime64[ns]
Open             float64
High             float64
Low              float64
Close            float64
Volume             int64
Symbol            object
dtype: object
        Date      Open      High       Low     Close  Volume Symbol
0 1985-01-02  1.664063  1.671875  1.664063  1.671875   17600    BMS
1 1985-01-03  1.656250  1.656250  1.640625  1.640625   27200    BMS
2 1985-01-07  1.640625  1.671875  1.640625  1.640625  414400    BMS
3 1985-01-08  1.640625  1.648438  1.632813  1.632813    9600    BMS
4 1985-01-09  1.625000  1.625000  1.609375  1.609375   16000    BMS


In [43]:
# Cast Volume as float to match datatype for same column in market_data DataFrame.
delisted_market_data['Volume'] = delisted_market_data['Volume'].astype('float64')
print(delisted_market_data.dtypes)

Date      datetime64[ns]
Open             float64
High             float64
Low              float64
Close            float64
Volume           float64
Symbol            object
dtype: object


In [44]:
# Similar to the data fetched from yfinance, the delisted price data will need to start at 1/1/2013.
delisted_market_data.drop(delisted_market_data[delisted_market_data['Date'] < '2013-01-01'].index, inplace=True)
delisted_market_data.reset_index(drop=True, inplace=True)
delisted_market_data.head()

,Date,Open,High,Low,Close,Volume,Symbol
0,2013-01-02,33.869999,34.099998,33.810001,34.000000,613400.0,BMS
1,2013-01-03,34.029999,34.130001,33.880001,33.950001,648900.0,BMS
2,2013-01-04,33.970001,34.200001,33.880001,34.060001,517400.0,BMS
3,2013-01-07,33.860001,34.040001,33.650002,34.020000,800900.0,BMS
4,2013-01-08,34.020000,34.200001,33.869999,33.930000,1045167.0,BMS


In [45]:
# Compare datatypes for the 2 DataFrames prior to concatenation.
print(delisted_market_data.dtypes)
print(market_data.dtypes)

Date      datetime64[ns]
Open             float64
High             float64
Low              float64
Close            float64
Volume           float64
Symbol            object
dtype: object
Date            datetime64[ns]
Open                   float64
High                   float64
Low                    float64
Close                  float64
Volume                 float64
Dividends              float64
Stock Splits           float64
Symbol                  object
dtype: object


In [46]:
# Concatenate the market_data and delisted_market_data DataFrames.
combined_market_data = pd.concat([market_data, delisted_market_data], ignore_index=True)
combined_market_data[['Dividends', 'Stock Splits']] = combined_market_data[['Dividends', 'Stock Splits']].fillna(0)
combined_market_data = combined_market_data.sort_values(['Symbol', 'Date']).reset_index(drop=True)
combined_market_data.head()

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits,Symbol
0,2013-01-02,34.919998,35.400002,34.099998,35.119999,13767900.0,0.0,0.0,ABBV
1,2013-01-03,35.000000,35.000000,34.160000,34.830002,16739300.0,0.0,0.0,ABBV
2,2013-01-04,34.619999,34.889999,34.250000,34.389999,21372100.0,0.0,0.0,ABBV
3,2013-01-07,34.150002,35.450001,34.150002,34.459999,17897100.0,0.0,0.0,ABBV
4,2013-01-08,34.290001,34.639999,33.360001,33.709999,17863300.0,0.0,0.0,ABBV


In [47]:
# Quick sanity check using delisted symbol SIAL
combined_market_data[combined_market_data['Symbol']=='SIAL'].head()

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits,Symbol
231046,2013-01-02,74.580002,75.190002,74.279999,74.769997,1134200.0,0.0,0.0,SIAL
231047,2013-01-03,74.980003,75.989998,74.570000,75.309998,518800.0,0.0,0.0,SIAL
231048,2013-01-04,75.510002,75.690002,75.160004,75.269997,468600.0,0.0,0.0,SIAL
231049,2013-01-07,74.580002,75.610001,74.540001,75.519997,401500.0,0.0,0.0,SIAL
231050,2013-01-08,75.110001,75.889999,75.040001,75.790001,486100.0,0.0,0.0,SIAL


Need to generate list of ticker symbols for stocks and NOBL.  My plan is to scrape dividend data from dividendhistory.org, and will need to convert BF-B to BF.B temporarily.

In [48]:
ticker_list = combined_market_data[~combined_market_data.Symbol.str.startswith(('^'))]['Symbol'].unique()
ticker_list = np.array(ticker_list, dtype=str)
ticker_list_for_div_hist = np.strings.replace(ticker_list, '-', '.')

# Quick inspection of tickers symbols to submit to dividendhistory.org
ticker_list_for_div_hist

array(['ABBV', 'ABT', 'ADM', 'ADP', 'AFL', 'ALB', 'AMCR', 'AOS', 'APD',
       'ATO', 'BDX', 'BEN', 'BF.B', 'BMS', 'BRO', 'CAH', 'CARR', 'CAT',
       'CB', 'CHD', 'CHRW', 'CINF', 'CL', 'CLX', 'CTAS', 'CVX', 'DOV',
       'ECL', 'ED', 'EMR', 'ERIE', 'ES', 'ESS', 'EXPD', 'FAST', 'FDO',
       'FDS', 'FRT', 'GD', 'GPC', 'GWW', 'HP', 'HRL', 'IBM', 'ITW', 'JNJ',
       'KMB', 'KO', 'KVUE', 'LEG', 'LIN', 'LOW', 'MCD', 'MDT', 'MKC',
       'MMM', 'NDSN', 'NEE', 'NOBL', 'NUE', 'O', 'OTIS', 'PBCT', 'PBI',
       'PEP', 'PG', 'PNR', 'PPG', 'RGLD', 'ROP', 'RTX', 'SHW', 'SIAL',
       'SJM', 'SPGI', 'SWK', 'SYK', 'SYY', 'T', 'TGT', 'TROW', 'VFC',
       'WBA', 'WMT', 'WST', 'XOM'], dtype='<U4')

In [49]:
# Check again for NaN, None, or NaT values.
ohlcv = ["Open", "High", "Low", "Close", "Volume"]
nan_data = combined_market_data[combined_market_data[ohlcv].isna().any(axis=1)]
print(len(nan_data))

0


In [50]:
# Need to check again for zero (0) values for Open, High, Low, Close, and Volume
# for stocks and NOBL.
ohlcv = ["Open", "High", "Low", "Close", "Volume"]
data_stocks= combined_market_data[~combined_market_data['Symbol'].str.startswith('^')]
zero_data_stocks = data_stocks[(data_stocks[ohlcv] == 0).any(axis=1)]
print(len(zero_data_stocks))

0


In [51]:
# Loop through the ticker symbols and fetch data from dividendhistory.org,
# and create div_hist_org DataFrame.

div_hist_org = pd.DataFrame()

for ticker in ticker_list_for_div_hist:
    url = f"https://dividendhistory.org/payout/{ticker}/"
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    }

    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")
        target_table = soup.find("table", {"id": "dividend-table"})

        if target_table:
            df = pd.read_html(StringIO(str(target_table)))[0]
            df.insert(0, "Symbol", ticker)
            div_hist_org = pd.concat([div_hist_org, df], ignore_index=True)
            print(f"[{ticker}] Data successfully scraped!")
        else:
            print(f"[{ticker}] Table with id 'dividend-table' not found on the page.")

    except requests.exceptions.RequestException as e:
        print(f"[{ticker}] Error fetching the webpage: {e}")
    except Exception as e:
        print(f"[{ticker}] An error occurred: {e}")

print(div_hist_org.head())

[ABBV] Data successfully scraped!
[ABT] Data successfully scraped!
[ADM] Data successfully scraped!
[ADP] Data successfully scraped!
[AFL] Data successfully scraped!
[ALB] Data successfully scraped!
[AMCR] Data successfully scraped!
[AOS] Data successfully scraped!
[APD] Data successfully scraped!
[ATO] Data successfully scraped!
[BDX] Data successfully scraped!
[BEN] Data successfully scraped!
[BF.B] Data successfully scraped!
[BMS] Data successfully scraped!
[BRO] Data successfully scraped!
[CAH] Data successfully scraped!
[CARR] Data successfully scraped!
[CAT] Data successfully scraped!
[CB] Data successfully scraped!
[CHD] Data successfully scraped!
[CHRW] Data successfully scraped!
[CINF] Data successfully scraped!
[CL] Data successfully scraped!
[CLX] Data successfully scraped!
[CTAS] Data successfully scraped!
[CVX] Data successfully scraped!
[DOV] Data successfully scraped!
[ECL] Data successfully scraped!
[ED] Data successfully scraped!
[EMR] Data successfully scraped!
[ERIE]

In [52]:
# Clean up the DataFrame to make it look presentable.
# Rename columns, subset to desired date frame, and clear
# out extraneous chars from numerical values.
div_hist_org = div_hist_org.rename(columns={'Ex-Dividend Date': 'Date',\
                                            'Cash Amount': 'Dividends',\
                                            '% Change': 'Comments'})
div_hist_org = div_hist_org[div_hist_org['Date'].between('2013-01-01', '2026-06-30')]
div_hist_org['Date'] = pd.to_datetime(div_hist_org['Date']).dt.normalize()
div_hist_org['Dividends'] = div_hist_org['Dividends'].str.replace('$', '', regex=False)
div_hist_org['Dividends'] = div_hist_org['Dividends'].astype(float)

div_hist_org = div_hist_org.sort_values(by=['Symbol', 'Date'])

print(len(div_hist_org))
print(div_hist_org.head(10))
print(div_hist_org.tail(10))

4454
   Symbol       Date Payout Date  Dividends Comments
56   ABBV 2013-01-11  2013-02-15       0.40      NaN
55   ABBV 2013-04-11  2013-05-15       0.40      NaN
54   ABBV 2013-07-11  2013-08-15       0.40      NaN
53   ABBV 2013-10-10  2013-11-15       0.40      NaN
52   ABBV 2014-01-13  2014-02-14       0.40      NaN
51   ABBV 2014-04-11  2014-05-15       0.42       5%
50   ABBV 2014-07-11  2014-08-15       0.42      NaN
49   ABBV 2014-10-10  2014-11-17       0.42      NaN
48   ABBV 2015-01-13  2015-02-13       0.49   16.67%
47   ABBV 2015-04-13  2015-05-15       0.51    4.08%
      Symbol       Date Payout Date  Dividends Comments
13316    XOM 2024-02-13  2024-03-11       0.95      NaN
13315    XOM 2024-05-14  2024-06-10       0.95      NaN
13314    XOM 2024-08-15  2024-09-10       0.95      NaN
13313    XOM 2024-11-14  2024-12-10       0.99    4.21%
13312    XOM 2025-02-12  2025-03-10       0.99      NaN
13311    XOM 2025-05-15  2025-06-10       0.99      NaN
13310    XOM 2025-08

In [53]:
# Cast the Payout Date column to datetime64[ns] so that is compatible with
# combined_market_data DataFrame.
div_hist_org['Payout Date'] = pd.to_datetime(div_hist_org['Payout Date']).dt.tz_localize(None)

In [54]:
# Compare datatypes of both DataFrames prior to merging.
print(combined_market_data.dtypes)
print(div_hist_org.dtypes)

Date            datetime64[ns]
Open                   float64
High                   float64
Low                    float64
Close                  float64
Volume                 float64
Dividends              float64
Stock Splits           float64
Symbol                  object
dtype: object
Symbol                 object
Date           datetime64[ns]
Payout Date    datetime64[ns]
Dividends             float64
Comments               object
dtype: object


In [55]:
# Need to put BF.B back to BF-B so I can merge the market_data and div_hist_org DataFrames.
div_hist_org['Symbol'] = div_hist_org['Symbol'].str.replace('.', '-', regex=False)
div_hist_org[div_hist_org['Symbol']=='BF-B'].head()

,Symbol,Date,Payout Date,Dividends,Comments
1817,BF-B,2013-03-06,2013-04-01,0.255,NaN
1816,BF-B,2013-06-03,2013-07-01,0.255,NaN
1815,BF-B,2013-09-03,2013-10-01,0.255,NaN
1814,BF-B,2013-12-02,2013-12-27,0.290,13.73%
1813,BF-B,2014-03-05,2014-04-01,0.290,NaN


In [56]:
# Before I forget, need to take a quick look at the splits for each symbol 
# that had splits, just so I have frame of reference.
combined_market_data[combined_market_data['Stock Splits'] > 0.0][['Date', 'Stock Splits', 'Symbol']].\
    sort_values(by='Symbol')

,Date,Stock Splits,Symbol
3393,2013-01-02,2.084202,ABT
10619,2014-10-01,1.139000,ADP
14883,2018-03-19,2.000000,AFL
22414,2026-01-15,0.200000,AMCR
22621,2013-05-16,2.000000,AOS
23476,2016-10-06,2.000000,AOS
26866,2016-10-03,1.081000,APD
35036,2022-04-01,1.025000,BDX
36003,2026-02-10,1.272000,BDX
36242,2013-07-26,3.000000,BEN


In [57]:
# Merge the combined_market_data and div_hist_org DataFrames.
merged = pd.merge(
    combined_market_data,
    div_hist_org,
    on=['Symbol', 'Date'],
    how='outer',
    suffixes=('_mktdata', '_org'),
    indicator=True
)

merged.head()

,Date,Open,High,Low,Close,Volume,Dividends_mktdata,Stock Splits,Symbol,Payout Date,Dividends_org,Comments,_merge
0,2013-01-02,34.919998,35.400002,34.099998,35.119999,13767900.0,0.0,0.0,ABBV,NaT,NaN,NaN,left_only
1,2013-01-03,35.000000,35.000000,34.160000,34.830002,16739300.0,0.0,0.0,ABBV,NaT,NaN,NaN,left_only
2,2013-01-04,34.619999,34.889999,34.250000,34.389999,21372100.0,0.0,0.0,ABBV,NaT,NaN,NaN,left_only
3,2013-01-07,34.150002,35.450001,34.150002,34.459999,17897100.0,0.0,0.0,ABBV,NaT,NaN,NaN,left_only
4,2013-01-08,34.290001,34.639999,33.360001,33.709999,17863300.0,0.0,0.0,ABBV,NaT,NaN,NaN,left_only


In [58]:
# Keep a golden dataset handy for backup since I will be doing lots of 
# reconciliation on the merged DataFrame in the subsequent cells.
# Create a .csv file for the same.
merged_golden_set = merged.copy()
merged_golden_set.to_csv('merged_golden_set.csv', index=False)

My first impression on the data from dividendhistory.org is that the dividends data is more robust and granular compared to the data from yfinance.  It should be noted that Google comments and discussions with Claude mention that yfinance data may be flighty at times.  Cursory examination of the data between the two services has uncovered the following dilemmas:

**Double split adjustments of dividends by yfinance:**
The dividend reported by yfinance shows an extra (and unnecessary) split adjustment.  The splits adjustment can be verified by taking the dividend from dividendhistory.org and dividing by the cumulative product of splits that occurred prior to that date.  The result of this calculation should be equal to the dividend from yfinance.

**Dividends declared as "special":**
A "special" dividend is a one-off dividend that is not part of the regularly scheduled dividend cycle for a company.  As such, it is preferable to have these dividends removed prior to the time series analysis.  The special dividend is usually paid on the same date as the regular dividend.  As an example, dividendhistory.org will show 0.25 regular and 0.235 special, but yfinance will show a single dividend of 0.485.  These amounts need to be reconciled (yfinance versus dividendhistory.org) and then the special dividend will ultimately be dropped.

**Duplicated payout dates by dividendhistory.org:**
For a small handful of dividends from dividendhistory.org, the Payout Date (and thus, the dividend amount) was duplicated.  These duplicated dividends will need to set to zero so that they will be omitted during the reconciliation process.

**Misaligned ex dividend dates between two services:**
The yfinance dividend and the dividendhistory.org dividend do not fall on the same date.  In all of these instances the dates are 1-3 days apart.  For purposes of reconciliation, the dividend from one service will need to be moved to the record for the dividend from the opposing service.

**Rounding of dividends by yfinance:**
In some cases the same dividend from yfinance versus dividendhistory.org is off by a fraction of a one cent, e.g., 0.248 from yfinance versus 0.2475 for dividendhistory.org.  For this example, I am inclined to go with the dividend from dividendhistory.org since it is more granular.

Comparing the dividends between both services will ensure I have all the necessary dividends for my time series machine learning, but at this time my overall inclination is for the dividendhistory.org data to be the higher source of truth for reconciliation, unless evidence shows otherwise.

In [59]:
# Look for instances in the yfinance dividends where the dividends where erroneously double-adjusted,
# such as the case for VFC's 1.062 stock split that occurred on 5/23/2019,
# or the 2.5 cumulative split for BF-B.

import pandas as pd
import numpy as np

def _find_qualifying_records(merged: pd.DataFrame) -> pd.DataFrame:
    """
    Core logic: identifies rows that qualify for the div_mktdata adjustment.
    Returns a DataFrame of qualifying records (empty if none), with columns:
    Symbol, Date, index (original row index in `merged`), Dividends_mktdata,
    Dividends_org, cum_split_factor, adjusted_org.
    """
    df = merged.sort_values(['Symbol', 'Date'])

    qualifying_records = []

    for symbol, group in df.groupby('Symbol', sort=False):
        splits_mask = group['Stock Splits'].notna() & (group['Stock Splits'] > 0)

        # Only process symbols that have at least one valid split
        if not splits_mask.any():
            continue

        group = group.sort_values('Date')
        split_events = group.loc[splits_mask, ['Date', 'Stock Splits']]

        # Rows eligible for comparison
        div_mask = (
            group['Dividends_mktdata'].notna() & (group['Dividends_mktdata'] > 0) &
            group['Dividends_org'].notna() & (group['Dividends_org'] > 0)
        )

        for idx, row in group.loc[div_mask].iterrows():
            dm = row['Dividends_mktdata']
            do = row['Dividends_org']

            # Skip rows where mktdata and org already agree
            if np.isclose(dm, do, atol=1e-6):
                continue

            # Nearest + all subsequent splits forward in time (cumulative product)
            fwd_splits = split_events.loc[split_events['Date'] >= row['Date'], 'Stock Splits']

            if fwd_splits.empty:
                continue  # no split forward in time, nothing to adjust against

            cum_split_factor = fwd_splits.prod()
            adjusted_org = do / cum_split_factor

            if np.isclose(adjusted_org, dm, atol=1e-6):
                qualifying_records.append({
                    'index': idx,
                    'Symbol': symbol,
                    'Date': row['Date'],
                    'Dividends_mktdata': dm,
                    'Dividends_org': do,
                    'cum_split_factor': cum_split_factor,
                    'adjusted_org': adjusted_org,
                })

    return pd.DataFrame(qualifying_records)


def print_qualifying_records(merged: pd.DataFrame) -> pd.DataFrame:
    """
    Identifies and prints records that qualify for the div_mktdata_adj adjustment.
    Does NOT modify `merged`. Returns the qualifying records DataFrame.
    """
    qual_df = _find_qualifying_records(merged)

    if qual_df.empty:
        print("\nNo records qualified for adjustment.")
    else:
        print(f"\n{len(qual_df)} record(s) qualify for adjustment:\n")
        print(qual_df.drop(columns='index').to_string(index=False))

    return qual_df


def adjust_mktdata_dividends(merged: pd.DataFrame) -> pd.DataFrame:
    """
    Creates 'div_mktdata_adj' column, initialized from 'Dividends_mktdata',
    then overwrites it with 'Dividends_org' for rows that qualify.
    """
    df = merged.copy()
    df['div_mktdata_adj'] = df['Dividends_mktdata']

    qual_df = _find_qualifying_records(df)

    if not qual_df.empty:
        df.loc[qual_df['index'], 'div_mktdata_adj'] = qual_df.set_index('index')['Dividends_org']

    return df


# --- Usage ---

# Print and inspect the dividends affected by the double adjustment.
qual_df = print_qualifying_records(merged)


191 record(s) qualify for adjustment:

Symbol       Date  Dividends_mktdata  Dividends_org  cum_split_factor  adjusted_org
   ADP 2013-03-06           0.381914          0.435           1.13900      0.381914
   ADP 2013-06-12           0.381914          0.435           1.13900      0.381914
   ADP 2013-09-11           0.381914          0.435           1.13900      0.381914
   ADP 2013-12-11           0.421422          0.480           1.13900      0.421422
   ADP 2014-03-12           0.421422          0.480           1.13900      0.421422
   ADP 2014-06-11           0.421422          0.480           1.13900      0.421422
   ADP 2014-09-10           0.421422          0.480           1.13900      0.421422
   APD 2013-03-27           0.656799          0.710           1.08100      0.656799
   APD 2013-06-27           0.656799          0.710           1.08100      0.656799
   APD 2013-09-27           0.656799          0.710           1.08100      0.656799
   APD 2013-12-30           0.656799

In [60]:
# Correct the records affected by the erroneous double-adjustment.
merged = adjust_mktdata_dividends(merged)

# Quick sanity check using ticker VFC.
merged[(merged['Symbol']=='VFC') & (merged['Date']=='2019-03-07')]\
    [['Date','Symbol','Dividends_mktdata','Dividends_org','div_mktdata_adj']]

,Date,Symbol,Dividends_mktdata,Dividends_org,div_mktdata_adj
260492,2019-03-07,VFC,0.480226,0.51,0.51


In [61]:
# This block of code will remove the records for special dividends.
import numpy as np

# Move records with 'Special' in Comments to special_div.
special_div = merged[merged['Comments'].str.contains('Special', case=False, na=False)].copy()

# Initialize the boolean column in merged.
# The special_div_check column will denote those dividends marked as special.
# The special_div_adj column will denote those rows where the sum of the special dividend and
# the ordinary dividend from dividendhistory.org yielded the dividend value returned by yfinance.
merged['special_div_check'] = False
merged['special_div_adj'] = False

# Set special_div_check to True in merged DataFrame for all rows with special dividends. 
merged['special_div_check'] = merged['Comments'].str.contains('Special', case=False, na=False)

# For each record in special_div, find matching Date-Symbol pairs in merged
# and check if special_div['Dividends_org'] + merged['Dividends_org'] == merged['Dividends_mktdata'].
for _, special_row in special_div.iterrows():
    mask = (merged['Date'] == special_row['Date']) & (merged['Symbol'] == special_row['Symbol']) &\
        (merged['Payout Date'] == special_row['Payout Date']) & (merged['special_div_check']==False)
    
    if mask.any():
        merged.loc[mask, 'special_div_adj'] = np.isclose(
            special_row['Dividends_org'] + merged.loc[mask, 'Dividends_org'],
            merged.loc[mask, 'div_mktdata_adj'],
            atol=1e-6
        )

# Print the output
print(merged[merged['special_div_adj']==True][['Date','Symbol','div_mktdata_adj','Dividends_org','special_div_check','special_div_adj']])

             Date Symbol  div_mktdata_adj  Dividends_org  special_div_check  \
36612  2014-12-29    BEN            0.650           0.15              False   
37430  2018-03-28    BEN            3.230           0.23              False   
134037 2022-11-14     HP            0.485           0.25              False   
134100 2023-02-13     HP            0.485           0.25              False   
134166 2023-05-17     HP            0.485           0.25              False   
134229 2023-08-16     HP            0.485           0.25              False   
134296 2023-11-17     HP            0.420           0.25              False   
134353 2024-02-12     HP            0.420           0.25              False   
134483 2024-08-16     HP            0.420           0.25              False   

        special_div_adj  
36612              True  
37430              True  
134037             True  
134100             True  
134166             True  
134229             True  
134296             True  
1

In [62]:
# Need to modify div_mktdata_adj to account for removal of special dividend.
merged.loc[merged['special_div_adj'], 'div_mktdata_adj'] = merged.loc[merged['special_div_adj'], 'Dividends_org']

# Check the output to verify equality between div_mktdata_adj and Dividends_org.
print(merged[merged['special_div_adj']==True][['Date','Symbol','div_mktdata_adj','Dividends_org','special_div_adj']])

             Date Symbol  div_mktdata_adj  Dividends_org  special_div_adj
36612  2014-12-29    BEN             0.15           0.15             True
37430  2018-03-28    BEN             0.23           0.23             True
134037 2022-11-14     HP             0.25           0.25             True
134100 2023-02-13     HP             0.25           0.25             True
134166 2023-05-17     HP             0.25           0.25             True
134229 2023-08-16     HP             0.25           0.25             True
134296 2023-11-17     HP             0.25           0.25             True
134353 2024-02-12     HP             0.25           0.25             True
134483 2024-08-16     HP             0.25           0.25             True


In [63]:
# Search for and return records with duplicate Payout Dates from Dividendhistory.org.
# Make sure to ignore those records already flagged for special dividends
# or special dividend adjustment.

duplicate_rows = (
    merged.dropna(subset=['Payout Date'])
    .pipe(lambda df: df[(df['special_div_check'] == False) & (df['special_div_adj'] == False)])
    .pipe(lambda df: df[df.duplicated(subset=['Symbol', 'Payout Date'], keep=False)])
    .sort_values(['Symbol', 'Payout Date'])[['Date','Symbol','Payout Date','div_mktdata_adj','Dividends_org']]
)
print(duplicate_rows)

             Date Symbol Payout Date  div_mktdata_adj  Dividends_org
25199  2023-07-28    AOS  2023-08-15            0.300         0.3000
25210  2023-08-14    AOS  2023-08-15            0.000         0.3000
26875  2016-09-29    APD  2016-11-14            0.000         0.8600
26880  2016-10-06    APD  2016-11-14            0.860         0.8600
49170  2017-12-29    CAH  2018-01-15            0.462         0.4624
49171  2018-01-01    CAH  2018-01-15              NaN         0.4624
254595 2022-09-15    TGT  2022-12-10            0.000         1.0800
254638 2022-11-15    TGT  2022-12-10            1.080         1.0800


In [64]:
# Mark those rows where div_mktdata_adj is either NaN or zero
set_for_zero = duplicate_rows[
    duplicate_rows['div_mktdata_adj'].isna() | (duplicate_rows['div_mktdata_adj'] == 0)
]
print(set_for_zero)

             Date Symbol Payout Date  div_mktdata_adj  Dividends_org
25210  2023-08-14    AOS  2023-08-15              0.0         0.3000
26875  2016-09-29    APD  2016-11-14              0.0         0.8600
49171  2018-01-01    CAH  2018-01-15              NaN         0.4624
254595 2022-09-15    TGT  2022-12-10              0.0         1.0800


In [65]:
# For those same records, set Dividends_org to zero in the merged Dataframe
merged.loc[set_for_zero.index, 'Dividends_org'] = 0

# Quick sanity check using symbol CAH.
merged[(merged['Symbol']=='CAH') & (merged['Payout Date']=='2018-01-15')]\
    [['Date','Symbol','Payout Date','div_mktdata_adj','Dividends_org']]

,Date,Symbol,Payout Date,div_mktdata_adj,Dividends_org
49170,2017-12-29,CAH,2018-01-15,0.462,0.4624
49171,2018-01-01,CAH,2018-01-15,NaN,0.0000


In [66]:
# Identify those Symbol-Date pairs that, for consecutive records, 
# have dividend from mktdata, but not from org, and vice versa.
# Ultimately, the dividend values from both services will need to be
# on the same row for purposes of reconciliation.

# NOTE:  The execution time for this cell is a little under 3 mins.

import pandas as pd
import numpy as np

df = merged.copy()

# Condition 1: div_mktdata_adj > 0, Payout Date is NaN, Dividends_org is NaN/0
condA = (
    (df['div_mktdata_adj'] > 0) &
    (df['Payout Date'].isna()) &
    (df['Dividends_org'].isna() | (df['Dividends_org'] == 0))
)

# Condition 2: div_mktdata_adj is NaN/0, Payout Date is not NaN, Dividends_org > 0
condB = (
    (df['div_mktdata_adj'].isna() | (df['div_mktdata_adj'] == 0)) &
    (df['Payout Date'].notna()) &
    (df['Dividends_org'] > 0)
)

df['_condA'] = condA
df['_condB'] = condB

pair_rows = []
pair_id = 0
pair_ids = pd.Series(np.nan, index=df.index)

for symbol, group in df.groupby('Symbol', sort=False):
    idx = group.index.to_list()
    for i in range(len(idx) - 1):
        i1, i2 = idx[i], idx[i + 1]
        r1, r2 = df.loc[i1], df.loc[i2]

        # Either (row1=A, row2=B) or (row1=B, row2=A)
        if (r1['_condA'] and r2['_condB']) or (r1['_condB'] and r2['_condA']):
            pair_id += 1
            pair_ids.loc[i1] = pair_id
            pair_ids.loc[i2] = pair_id
            pair_rows.extend([i1, i2])

result = df.loc[pair_rows].copy()
result['pair_id'] = pair_ids.loc[pair_rows].astype(int)
result = result.drop(columns=['_condA', '_condB'])

result[['Date','Symbol','div_mktdata_adj','Payout Date','Dividends_org']]

,Date,Symbol,div_mktdata_adj,Payout Date,Dividends_org
4469,2017-04-11,ABT,0.265,NaT,NaN
4470,2017-04-12,ABT,0.000,2017-05-15,0.2650
40998,2018-12-04,BF-B,0.000,2019-01-02,0.1660
40999,2018-12-06,BF-B,0.166,NaT,NaN
55537,2023-07-19,CAT,1.300,NaT,NaN
55538,2023-07-20,CAT,0.000,2023-08-18,1.3000
98442,2018-10-05,ERIE,0.840,NaT,NaN
98443,2018-10-08,ERIE,0.000,2018-10-23,0.8400
104279,2014-12-30,ESS,1.300,NaT,NaN
104280,2014-12-31,ESS,0.000,2015-01-15,1.3000


In [67]:
# For each affected Symbol-Date pair, copy Dividends_org value to record with
# non-NaN div_mktdata_adj value.  Then in the source record set Dividends_org to zero.
for pid, grp in result.groupby('pair_id'):
    idx1, idx2 = grp.index[0], grp.index[1]

    val1 = merged.loc[idx1, 'Dividends_org']
    val2 = merged.loc[idx2, 'Dividends_org']

    # Identify which one actually holds the non-NaN (and non-zero) value.
    if pd.notna(val1) and val1 != 0:
        merged.loc[idx2, 'Dividends_org'] = val1
        merged.loc[idx1, 'Dividends_org'] = 0
    elif pd.notna(val2) and val2 != 0:
        merged.loc[idx1, 'Dividends_org'] = val2
        merged.loc[idx2, 'Dividends_org'] = 0

# Quick sanity check for symbol HRL.  Need to make sure that same record holds
# both the div_mktdata_adj and Dividends_org values.
merged[(merged['Symbol']=='HRL') & (merged['Date'].between('2018-01-11','2018-01-12'))]\
[['Date','Symbol','div_mktdata_adj','Payout Date','Dividends_org']]

,Date,Symbol,div_mktdata_adj,Payout Date,Dividends_org
136218,2018-01-11,HRL,0.188,NaT,0.1875
136219,2018-01-12,HRL,0.000,2018-02-15,0.0000


In [68]:
# Last round of reconciliations are coming up.
# I'll create another golden set just to be safe.
merged_adjusted_set = merged.copy()
merged_adjusted_set.to_csv('merged_adjusted_set.csv', index=False)

In [69]:
# Now it's time to compare apples to apples for the dividends that originated
# from yfinance versus those from dividendhistory.org.
# I'll be creating four DataFrames.

# Dividends that are either equal or off by 0.0005 or less between the two services.
# These dividends I will consider fully reconciled and will leave untouched.
matched = merged[
    (merged['div_mktdata_adj'].notna()) &
	(merged['div_mktdata_adj'] != 0) &
	(merged['Dividends_org'].notna()) &
	(merged['Dividends_org'] != 0) &
    (merged['special_div_check'] == False) &
    ((merged['div_mktdata_adj'] - merged['Dividends_org']).abs() <= 0.0005 + 1e-9)
]

# Dividends that are either equal or off by 0.0005 or less between the two services.
# These might be considered rounding errors, but close inspection will be required.
mismatched = merged[
    (merged['div_mktdata_adj'].notna()) &
	(merged['div_mktdata_adj'] != 0) &
	(merged['Dividends_org'].notna()) &
	(merged['Dividends_org'] != 0) &
    (merged['special_div_check'] == False) &
    ((merged['div_mktdata_adj'] - merged['Dividends_org']).abs() > 0.0005 + 1e-9)
]

# Records only in market_data (missing from div_hist_org)
only_in_mktdata = merged[
    (merged['div_mktdata_adj'].notna()) &
	(merged['div_mktdata_adj'] != 0) &
    (merged['special_div_check'] == False) &
	((merged['Dividends_org'].isna()) |
	(merged['Dividends_org'] == 0))
]

# Records only in div_hist_org (missing from market_data)
only_in_org = merged[
    (merged['Dividends_org'].notna()) &
	(merged['Dividends_org'] != 0) &
    (merged['special_div_check'] == False) &
	((merged['div_mktdata_adj'].isna()) |
	(merged['div_mktdata_adj'] == 0))
]

print(f"Found {len(matched)} matched records:")
print(matched[['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])

print(f"Found {len(mismatched)} mismatched records:")
print(mismatched[['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])

print(f"\nFound {len(only_in_mktdata)} records in market_data but not in div_hist_org:")
print(only_in_mktdata[['Symbol', 'Date', 'div_mktdata_adj']])

print(f"\nFound {len(only_in_org)} records in div_hist_org but not in market_data:")
print(only_in_org[['Symbol', 'Date', 'Dividends_org']])


Found 4144 matched records:
       Symbol       Date  div_mktdata_adj  Dividends_org
7        ABBV 2013-01-11             0.40           0.40
68       ABBV 2013-04-11             0.40           0.40
131      ABBV 2013-07-11             0.40           0.40
195      ABBV 2013-10-10             0.40           0.40
259      ABBV 2014-01-13             0.40           0.40
...       ...        ...              ...            ...
275411    XOM 2025-05-15             0.99           0.99
275474    XOM 2025-08-15             0.99           0.99
275538    XOM 2025-11-14             1.03           1.03
275598    XOM 2026-02-12             1.03           1.03
275662    XOM 2026-05-15             1.03           1.03

[4144 rows x 4 columns]
Found 122 mismatched records:
       Symbol       Date  div_mktdata_adj  Dividends_org
21134    AMCR 2020-11-23         0.590000         0.5875
21195    AMCR 2021-02-23         0.590000         0.5875
21259    AMCR 2021-05-25         0.590000         0.5875
21331

In [70]:
# It will be easiest to start with the only_in_org DataFrame.
# Check for the ticker symbols that are affected.
only_in_org.groupby('Symbol').size()

Symbol
AMCR    13
BMS     26
CB      14
CHRW     1
EXPD     1
FDO      8
GPC      1
KMB      1
LIN      1
MCD      1
MDT      1
NEE      1
PBCT    37
PEP      1
SIAL    11
WBA     48
dtype: int64

In [71]:
# Quick check of those symbols with only one record in the output.
print(only_in_org[only_in_org['Symbol']=='CHRW']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_org[only_in_org['Symbol']=='EXPD']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_org[only_in_org['Symbol']=='LIN']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_org[only_in_org['Symbol']=='KMB']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_org[only_in_org['Symbol']=='MDT']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_org[only_in_org['Symbol']=='PEP']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])

      Symbol       Date  div_mktdata_adj  Dividends_org
66438   CHRW 2026-06-05              0.0           0.63
       Symbol       Date  div_mktdata_adj  Dividends_org
110542   EXPD 2026-06-01              0.0           0.81
       Symbol       Date  div_mktdata_adj  Dividends_org
162869    LIN 2026-06-04              0.0            1.6
       Symbol       Date  div_mktdata_adj  Dividends_org
151900    KMB 2026-06-05              0.0           1.28
       Symbol       Date  div_mktdata_adj  Dividends_org
171119    MDT 2018-09-27              0.0            0.5
       Symbol       Date  div_mktdata_adj  Dividends_org
207300    PEP 2026-06-05              0.0           1.48


For symbols BMS, FDO, PBCT, SIAL, and WBA, the returned counts make sense as dividendhistory.org is the only free online source of historical dividends for delisted stocks, and yfinance does not have this data. For the symbols with count=1, I had checked each individually, and yfinance had strangely omitted the June 2026 dividend for these. For this set of cases I will go with the dividends reported by dividendhistory.org.  For symbol MDT the 9/27/2018 dividend was omitted by yfinance.  Again, I'll defer to the dividend reported by dividendhistory.org for this symbol.  Regarding AMCR, earlier in this notebook I had to drop 1200+ records for AMCR due to zeroed-out Volume. The dividends in the only_in_org DataFrame for AMCR are for 2014-2019 and are devoid of Open, High, Low, Close, and Volume data. Since AMCR was only added to the index in 2020, I plan on dropping these records.  They will naturally be dropped when I do a final sweep of records that have NaN/zero values.  Finally for symbol CB, in January 2016 the old Chubb company merged with ACE Limited. After the merger ACE assumed the CB ticker for the new combined entity under the name Chubb Limited. While doing my own spot checking, I noticed that the yfinance dividends for CB have different dates and amounts than those from dividendhistory.org. Further research showed that dividendhistory.org's data is for the old Chubb lineage which is what I desire for the time series machine learning since the old Chubb was part of the Dividend Aristocrats index up to 2016. In light of this fact the dividendhistory.org data will be the source of truth for CB and will be copied over to the div_mktdata_adj column in the next cell.

In [72]:
# For the records in the only_in_org DataFrame, copy the Dividends_org value 
# to div_mktdata_adj in the merged DataFrame.

idx_to_update = only_in_org.index
merged.loc[idx_to_update, 'div_mktdata_adj'] = merged.loc[idx_to_update, 'Dividends_org']

# Recalculate the only_in_org DataFrame and inspect the results.
only_in_org = merged[
    (merged['Dividends_org'].notna()) &
	(merged['Dividends_org'] != 0) &
    (merged['special_div_check'] == False) &
	((merged['div_mktdata_adj'].isna()) |
	(merged['div_mktdata_adj'] == 0))
]

# Quick sanity check
print(f"\nFound {len(only_in_org)} records in div_hist_org but not in market_data:")
print(only_in_org[['Symbol', 'Date', 'Dividends_org']])


Found 0 records in div_hist_org but not in market_data:
Empty DataFrame
Columns: [Symbol, Date, Dividends_org]
Index: []


In [73]:
# Next in line is the only_in_mktdata DataFrame.  List the symbols affected.
only_in_mktdata.groupby('Symbol').size()

Symbol
CB      12
NOBL     2
dtype: int64

In [74]:
print(only_in_mktdata[only_in_mktdata['Symbol']=='CB']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])
print(only_in_mktdata[only_in_mktdata['Symbol']=='NOBL']\
    [['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])

      Symbol       Date  div_mktdata_adj  Dividends_org
56334     CB 2013-03-26             0.49            NaN
56414     CB 2013-07-19             0.51            NaN
56462     CB 2013-09-26             0.51            NaN
56536     CB 2014-01-13             0.63            NaN
56586     CB 2014-03-26             0.63            NaN
56666     CB 2014-07-21             0.65            NaN
56714     CB 2014-09-26             0.65            NaN
56769     CB 2014-12-15             0.65            NaN
56839     CB 2015-03-27             0.65            NaN
56902     CB 2015-06-26             0.67            NaN
56966     CB 2015-09-28             0.67            NaN
57030     CB 2015-12-29             0.67            NaN
       Symbol       Date  div_mktdata_adj  Dividends_org
186691   NOBL 2013-12-24            0.067            NaN
186753   NOBL 2014-03-26            0.090            NaN


For CB the situation is as described in one of a previous cell.  The final corrective action for CB will be performed in the next cell.  For NOBL, yfinance has dividends for ex div dates 12/24/2013 and 3/26/2014, but dividendhistory.org does not.  With Claude's help I was able to determine that these dividends are valid, thus the values in div_mktdata_adj are accurate, and I will leave these 2 records for NOBL untouched.

In [75]:
# Since the only_in_mktdata records for CB are from the ACE lineage, these need to be set to zero.
mask = only_in_mktdata['Symbol'] == 'CB'
idx_to_update = only_in_mktdata[mask].index

merged.loc[idx_to_update, 'div_mktdata_adj'] = 0

# Recalculate the only_in_mktdata DataFrame and inspect the results.
# Only those records for NOBL should remain.
only_in_mktdata = merged[
    (merged['div_mktdata_adj'].notna()) &
	(merged['div_mktdata_adj'] != 0) &
    (merged['special_div_check'] == False) &
	((merged['Dividends_org'].isna()) |
	(merged['Dividends_org'] == 0))
]

print(f"\nFound {len(only_in_mktdata)} records in market_data but not in div_hist_org:")
print(only_in_mktdata[['Symbol', 'Date', 'div_mktdata_adj']])


Found 2 records in market_data but not in div_hist_org:
       Symbol       Date  div_mktdata_adj
186691   NOBL 2013-12-24            0.067
186753   NOBL 2014-03-26            0.090


In [76]:
# Finally we have the mismatched DataFrame.  List the symbols affected.
mismatched.groupby('Symbol').size()

Symbol
AMCR    11
BF-B     4
CTAS     1
HP       1
O       44
SPGI    54
VFC      7
dtype: int64

In [77]:
# I suspect O and SPGI are going to be special cases.  
# Do a quick rundown on the other symbols for now.
print(mismatched[mismatched['Symbol']=='AMCR'][['Symbol', 'Date', 'div_mktdata_adj',\
                                                'Dividends_org','special_div_check','special_div_adj']])
print(mismatched[mismatched['Symbol']=='BF-B'][['Symbol', 'Date', 'div_mktdata_adj',\
                                                'Dividends_org','special_div_check','special_div_adj']])
print(mismatched[mismatched['Symbol']=='CTAS'][['Symbol', 'Date', 'div_mktdata_adj',\
                                                'Dividends_org','special_div_check','special_div_adj']])
print(mismatched[mismatched['Symbol']=='HP'][['Symbol', 'Date', 'div_mktdata_adj',\
                                                'Dividends_org','special_div_check','special_div_adj']])
print(mismatched[mismatched['Symbol']=='VFC'][['Symbol', 'Date', 'div_mktdata_adj',\
                                                'Dividends_org','special_div_check','special_div_adj']])

      Symbol       Date  div_mktdata_adj  Dividends_org  special_div_check  \
21134   AMCR 2020-11-23            0.590         0.5875              False   
21195   AMCR 2021-02-23            0.590         0.5875              False   
21259   AMCR 2021-05-25            0.590         0.5875              False   
21331   AMCR 2021-09-07            0.590         0.5875              False   
21637   AMCR 2022-11-22            0.615         0.6125              False   
21702   AMCR 2023-02-28            0.615         0.6125              False   
21761   AMCR 2023-05-23            0.615         0.6125              False   
21833   AMCR 2023-09-06            0.615         0.6125              False   
22139   AMCR 2024-11-21            0.640         0.6375              False   
22202   AMCR 2025-02-26            0.640         0.6375              False   
22262   AMCR 2025-05-22            0.640         0.6375              False   

       special_div_adj  
21134            False  
21195        

My individual investigations yielded the following.  The AMCR records are an issue with rounding by yfinance (rounding was performed to the nearest quarter of 1 cent), and I am content with going with the dividendhistory.org version since is more granular.  For symbols BF-B, CTAS, and VFC the individual discrepancies cannot be attributed to split adjustments, rounding, or special dividends.  I'll be deferring to the data from dividendhistory.org for these.  For symbol HP yfinance reported a special dividend that was not reported by dividendhistory.org.  A quick investigation showed that in this case yfinance is correct.  I will need to manually change the div_mktdata_adj value for this record accordingly.

My research shows that SPGI split 1057 for 1000 on 7/1/2026.  Yfinance shows the adjusted dividends correctly, but dividendhistory.org does not.  I've emailed the support team at dividendhistory.org regarding the need for the split adjustment, but am still waiting on a response.  At this time, the values in div_mktdata_adj are accurate, and I will leave this data untouched.

In November 2021, Realty Income (O) merged with VEREIT.  Both services are using different adjustment factors based on the merger, but since each of the differences are so narrow I'm inclined to use the dividends from dividendhistory.com just for the sake of consistency.

In [79]:
# Manually correct the dividend data for HP for the 5/16/2024 date.
mask = (merged['Symbol'] == 'HP') & (merged['Date'] == pd.Timestamp('2024-05-16'))
merged.loc[mask, 'div_mktdata_adj'] = 0.25

print(merged[mask])

             Date       Open       High    Low      Close    Volume  \
134420 2024-05-16  38.060001  38.380001  37.75  38.209999  874300.0   

        Dividends_mktdata  Stock Splits Symbol Payout Date  Dividends_org  \
134420               0.42           0.0     HP  2024-05-31           0.25   

       Comments _merge  div_mktdata_adj  special_div_check  special_div_adj  
134420      NaN   both             0.25              False            False  


In [80]:
# For the remaining symbols (aside from SPGI) in the mismatched DataFrame, 
# copy the Dividends_org value to div_mktdata_adj in the merged DataFrame.
mask = mismatched['Symbol'].isin(['AMCR', 'BF-B', 'CTAS', 'O', 'VFC'])
idx_to_update = mismatched[mask].index

merged.loc[idx_to_update, 'div_mktdata_adj'] = merged.loc[idx_to_update, 'Dividends_org']

# Recalculate the mismatched DataFrame and inspect the results.
# Only the records for SPGI should remain.
mismatched = merged[
    (merged['div_mktdata_adj'].notna()) &
	(merged['div_mktdata_adj'] != 0) &
	(merged['Dividends_org'].notna()) &
	(merged['Dividends_org'] != 0) &
    (merged['special_div_check'] == False) &
    ((merged['div_mktdata_adj'] - merged['Dividends_org']).abs() > 0.0005 + 1e-9)
]

print(f"Found {len(mismatched)} mismatched records:")
print(mismatched[['Symbol', 'Date', 'div_mktdata_adj', 'Dividends_org']])

Found 54 mismatched records:
       Symbol       Date  div_mktdata_adj  Dividends_org
235222   SPGI 2013-02-22         0.264901           0.28
235286   SPGI 2013-05-24         0.264901           0.28
235349   SPGI 2013-08-23         0.264901           0.28
235413   SPGI 2013-11-22         0.264901           0.28
235474   SPGI 2014-02-24         0.283822           0.30
235537   SPGI 2014-05-23         0.283822           0.30
235600   SPGI 2014-08-22         0.283822           0.30
235664   SPGI 2014-11-21         0.283822           0.30
235725   SPGI 2015-02-23         0.312204           0.33
235788   SPGI 2015-05-22         0.312204           0.33
235852   SPGI 2015-08-24         0.312204           0.33
235916   SPGI 2015-11-23         0.312204           0.33
235977   SPGI 2016-02-23         0.340587           0.36
236041   SPGI 2016-05-24         0.340587           0.36
236105   SPGI 2016-08-24         0.340587           0.36
236169   SPGI 2016-11-23         0.340587           0.36
23

In [81]:
# Check again for zero or NaN values for Open, High, Low, Close, and Volume for stocks
# and NOBL.

ohlcv = ["Open", "High", "Low", "Close", "Volume"]
data_stocks = merged[~merged['Symbol'].str.startswith('^')]
zero_data_stocks = data_stocks[(data_stocks[ohlcv] == 0).any(axis=1) | data_stocks[ohlcv].isna().any(axis=1)]
print(len(zero_data_stocks))

13


In [82]:
zero_data_stocks

,Date,Open,High,Low,Close,Volume,Dividends_mktdata,Stock Splits,Symbol,Payout Date,Dividends_org,Comments,_merge,div_mktdata_adj,special_div_check,special_div_adj
20518,2014-02-25,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2014-03-18,0.880,NaN,right_only,0.880,False,False
20529,2014-09-02,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2014-09-23,1.095,24.43%,right_only,1.095,False,False
20553,2015-09-08,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2015-09-29,0.985,-10.05%,right_only,0.985,False,False
20567,2016-02-23,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2016-03-15,0.965,-2.03%,right_only,0.965,False,False
20581,2016-09-06,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2016-09-27,1.100,13.99%,right_only,1.100,False,False
20589,2017-02-24,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2017-03-17,0.990,-10%,right_only,0.990,False,False
20619,2017-09-05,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2017-09-26,1.190,20.2%,right_only,1.190,False,False
20640,2018-02-27,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2018-03-20,1.030,-13.45%,right_only,1.030,False,False
20668,2018-09-07,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2018-09-28,1.175,14.08%,right_only,1.175,False,False
20732,2019-03-04,NaN,NaN,NaN,NaN,NaN,NaN,NaN,AMCR,2019-03-25,1.055,-10.21%,right_only,1.055,False,False


In the cases of CAH and LOW these were dates served up by dividendhistory.org and were market holidays.  The dividends for these records have since been reconciled.  In the case of AMCR, unfortunately my Norgate data for this symbol only extends back to early 2019.  Thankfully this is immaterial since AMCR was added to the Dividend Aristocrats in 2020.  In light of the above information, I will drop these records from the merged DataFrame and then reset index.

In [83]:
idx_to_update = zero_data_stocks.index

merged.drop(zero_data_stocks.index, inplace=True)
merged.reset_index(drop=True, inplace=True) 

# Sanity check
ohlcv = ["Open", "High", "Low", "Close", "Volume"]
data_stocks = merged[~merged['Symbol'].str.startswith('^')]
zero_data_stocks = data_stocks[(data_stocks[ohlcv] == 0).any(axis=1) | data_stocks[ohlcv].isna().any(axis=1)]
print(len(zero_data_stocks))

0


In [84]:
# Check zero (0) values for Open, High, Low, and Close for the indices.
# Zero volume is immaterial since I only need price data for the
# indices for my analysis.
ohlc = ["Open", "High", "Low", "Close"]
data_indices = merged[merged['Symbol'].str.startswith('^')]
zero_data_indices = data_indices[(data_indices[ohlc] == 0).any(axis=1)]
print(len(zero_data_indices))

0


In [85]:
# Check for the existence of duped records from the subset of non-special dividends

special_divs_removed = merged[merged['special_div_check']==False]
duplicate_count = special_divs_removed.duplicated(subset=['Date', 'Symbol']).sum()
print(duplicate_count)

0


In [86]:
# Drop the following columns since they are no longer needed:  Dividends_mktdata,
# Stock Splits, Payout Date, Dividends_org, Comments, _merge, special_div_check, 
# special_div_adj.
merged = merged.drop(columns=['Dividends_mktdata', 'Stock Splits', 'Payout Date', 'Dividends_org',\
                               'Comments', '_merge', 'special_div_check', 'special_div_adj'])
merged.rename(columns={'div_mktdata_adj': 'Dividend'}, inplace=True)
merged.head()

,Date,Open,High,Low,Close,Volume,Symbol,Dividend
0,2013-01-02,34.919998,35.400002,34.099998,35.119999,13767900.0,ABBV,0.0
1,2013-01-03,35.000000,35.000000,34.160000,34.830002,16739300.0,ABBV,0.0
2,2013-01-04,34.619999,34.889999,34.250000,34.389999,21372100.0,ABBV,0.0
3,2013-01-07,34.150002,35.450001,34.150002,34.459999,17897100.0,ABBV,0.0
4,2013-01-08,34.290001,34.639999,33.360001,33.709999,17863300.0,ABBV,0.0


In [87]:
# Save 3 DataFrames to respective .csv files
merged.to_csv('merged_final_for_time_series.csv', index=False)
aristocrats_complete.to_csv('aristocrats_complete_for_time_series.csv', index=False)
indices.to_csv('indices_for_time_series.csv', index=False)

Some final assumptions about the above data wrangling need to be noted.  It is assumed that the price and volume data from yfinance is accurate and is adjusted only for splits and not for dividends.  It is also assumed that the dividend data from dividendhistory.org is accurate (that is, within measure, since no service is perfect) and is successfully adjusted for splits.